// @animations/RotatingLogo/RotatingLogo.js
// @flow

import React from 'react';
import { TTLogoPath } from '@components/Logo/TTLogo';

const styles = {
    container: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'flex-start',
    },
    svg: {
        height: '30px',
        width: '30px',
    },
    text: {
        textTransform: 'uppercase',
        fontSize: '25px',
        fontWeight: 100,
        color: 'white',
        lineHeight: '25px',
        margin: 0,
        padding: '4px 0 0 10px',
    },
};

/**
 * RotatingLogo
 */
export const RotatingLogo = () => (
    <div style={styles.container}>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 140 140" style={styles.svg}>
            <g fill="red" transform="translate(20,20)">
                <TTLogoPath />
                <animateTransform
                    attributeType="xml"
                    attributeName="transform"
                    type="rotate"
                    from="0 50 50"
                    to="360 50 50"
                    dur="2s"
                    additive="sum"
                    repeatCount="indefinite"
                />
            </g>
        </svg>
        <p style={styles.text}>LOADING</p>
    </div>
);

export default RotatingLogo;
