// @data/Constants/config.js
/* eslint-disable no-console */
export const HOST = global.HOST || 'http://localhost:8000/';
export const API_URI = global.API_URL || `${HOST}api/`;

let APP_VERSION = false;
if (process && process.env && process.env.REACT_APP_VERSION) {
    APP_VERSION = process.env.REACT_APP_VERSION;
}

let APP_GIT = false;
if (process && process.env && process.env.REACT_APP_GIT && process.env.REACT_APP_GIT !== 'null') {
    APP_GIT = process.env.REACT_APP_GIT;
}

if (!process) {
    // dont log in tests
    console.info('APP_VERSION', APP_VERSION);
    console.info('APP_GIT', APP_GIT);
    console.info('HOST', HOST);
    console.info('API_URI', API_URI);
}

export const DEFAULT_FETCH_CONFIG = {
    headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
    },
    mode: 'cors',
    cache: 'default',
};

export const GET = {
    method: 'GET',
    ...DEFAULT_FETCH_CONFIG,
};

export const POST = {
    method: 'POST',
    ...DEFAULT_FETCH_CONFIG,
};

export const PUT = {
    method: 'PUT',
    ...DEFAULT_FETCH_CONFIG,
};

export const DELETE = {
    method: 'DELETE',
    ...DEFAULT_FETCH_CONFIG,
};

export const client = {
    get: uri => fetch(`${API_URI}${uri}`, GET).then(response => response.json()),
    post: (uri, body) => fetch(`${API_URI}${uri}`, { ...POST, body: JSON.stringify(body) }).then(response => response.json()),
    put: (uri, body) => fetch(`${API_URI}${uri}`, { ...PUT, body: JSON.stringify(body) }).then(response => response.json()),
    delete: uri => fetch(`${API_URI}${uri}`, DELETE),
};

export default {
    API_URI, APP_VERSION, APP_GIT, client,
};
