// @components/PrivateRoute/PrivateRoute.js
// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { is } from 'ramda';
import { Redirect } from 'react-router';
import {
    withRouter,
    Route,
} from 'react-router-dom';

type TPrivateRouteProps = {
    component: Component,
    path: string,
    validation: Function,
    history: any,
    isAuthenticated: boolean,
};

export const PrivateRoute = (props: TPrivateRouteProps) => {
    const {
        component, path, validation, history, isAuthenticated,
    } = props;
    // on /auth do nothing
    if (history.location.pathname === '/auth') return null;
    if (!isAuthenticated) return (<Redirect push to="/auth" />);
    const valid = validation(props, path);
    if (is(String, valid)) {
        history.push(valid);
        return null;
    }

    const Element = component;

    return (<Route
        path={path}
        {...props}
        render={p => <Element {...p} />}
    />);
};

const mapStateToProps = ({ auth }, { history }) => ({
    isAuthenticated: auth.isAuthenticated,
    user: auth.user,
    history,
});

export default withRouter(connect(mapStateToProps)(PrivateRoute));
