// @data/Resources/ResourcesTypes.js
// @flow
export type DurationType = {
    hours?: number,
    minutes?: number,
};

// general DateType if we want to change it afterwards
export type DateType = string; // 2017-09-25T00:00:00.000Z

export type ApiUserType = {
    usr_id: number,
    usr_firstname: string,
    usr_lastname: string,
    usr_employment_start: string | null,
    usr_employment_stop: string | null,
};

export type ApiPeriodTypeType = {
    pty_config: {
        icon: string,
        bgcolor: string,
        color: string,
        types: {
            period: boolean,
            fullday: boolean,
            halfday: boolean,
            duration: boolean,
        },
    },
    pty_id: string,
    pty_name: string,
};

export type ApiPeriodType = {
    per_break: DurationType,
    per_comment: string | null,
    per_day_id: number,
    per_duration: DurationType,
    per_id: number,
    per_pty_id: string, // TODO "Work"
    per_start: DurationType,
    per_stop: DurationType,
    pty_name: string,
};

export type ApiDayType = {
    day_date: DateType,
    day_id: number,
    day_target_time: DurationType,
    day_usr_id: number,
    periods: Array<ApiPeriodType>,
    remaining: DurationType,
};

export type ApiTimeSheetType = {
    carryFrom: DateType,
    carryTo: DateType,
    carryTime: DurationType,
    days: Array<ApiDayType>,
};
