// @components/LoadingApplication/LoadingApplication.js
// @flow

import React from 'react';

import { RotatingLogo } from '@animations';

import './LoadingApplication.css';

const styles = {
    container: {
        display: 'flex',
        flexDirection: 'column',
        // width: '100%',
        height: '100%',
    },
    topContainer: {
        backgroundColor: '#434a54',
        borderBottom: '2px solid rgb(46, 51, 58)',
        flex: 1,
        position: 'relative',
        transition: 'top 0.5s ease-in-out',
    },
    bottomContainer: {
        backgroundColor: '#434a54',
        borderTop: '2px solid rgb(46, 51, 58)',
        flex: 1,
        transition: 'bottom 0.5s ease-in-out',
    },
    logoContainer: {
        background: 'rgb(46, 51, 58)',
        position: 'absolute',
        bottom: '-26px',
        left: '50%',
        padding: '10px 20px',
        borderRadius: '15px',
        marginLeft: '-65px',
        border: '1px solid rgb(37, 41, 47)',
    },
};

/**
 * LoadingApplication
 */
export const LoadingApplication = () => (
    <div style={styles.container}>
        <div style={styles.topContainer}>
            <div style={styles.logoContainer}>
                <RotatingLogo />
            </div>
        </div>
        <div style={styles.bottomContainer} />
    </div>
);

export default LoadingApplication;
